### PR TITLE
Iox #1440 fix file lock leak in release 2.0

### LINKED ISSUE #1440: sometimes the app has FD leak, when the APP use fileLock
## Required information

**Operating system:**
E.g. Ubuntu 18.04 LTS

**Compiler version:**
E.g. GCC 7.4.0

**Observed result or behaviour:**
There are some FD leakage, it could easily reach to the max opened FDs of a process in Linux platform.

**Expected result or behaviour:**
no FD leakage.

**Conditions where it occurred / Performed steps:**

APP_A
```
int main()
{
    //sending();

    iox::runtime::PoshRuntime::initRuntime(APP_NAME);
    auto maybeFileLock = iox::posix::FileLock::create(TEST_NAME);


    while(!iox::posix::hasTerminationRequested()) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return (EXIT_SUCCESS);
}
```

APP_B

```
int main()
{
    iox::runtime::PoshRuntime::initRuntime(APP_NAME);

    constexpr char TEST_NAME[] = "TestProcessName0630";

    for (int i = 0; i < 1024; i++)
    {
        auto maybeFileLock = iox::posix::FileLock::create(TEST_NAME);

        if (!maybeFileLock.has_error()) {
            // process has exit
            std::cout << "publisher has exit" << std::endl;
        } else {
            if (maybeFileLock.get_error() == iox::posix::FileLockError::LOCKED_BY_OTHER_PROCESS) {
                std::cout << "file is locked by another process" << std::endl;
            } else {
                // normally, this shouldn't happend
                std::cout << "Error occured while acquiring file lock named " << std::endl;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

{
        auto maybeFileLock = iox::posix::FileLock::create(TEST_NAME);

        if (!maybeFileLock.has_error()) {
            // process has exit
            std::cout << "publisher has exit" << std::endl;
        } else {
            if (maybeFileLock.get_error() == iox::posix::FileLockError::LOCKED_BY_OTHER_PROCESS) {
                std::cout << "file is locked by another process" << std::endl;
            } else {
                // normally, this shouldn't happend
                std::cout << "Error occured while acquiring file lock named " << std::endl;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
}


    return (EXIT_SUCCESS);
}
```

and start APP_A first, then start APP_B.
You will see this issue.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
This bug occurs only in 2.0.x . On master the `FileLock` was refactored and with that the bug is already solved.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1440 and backports #1338 to 2.0

### CODE REVIEW

**Comment by elBoberido** [comment]
@elfenpiff can you also backport #1338 if necessary?

**Comment by elBoberido** [comment]
@elfenpiff could you please also add #1338 to the references of the PR description

**Comment by dkroenke** [comment]
On hold until a 2.0.3 bugfix release is agreed
- [ ] release is confirmed and agreed

File: iceoryx_hoofs/source/posix_wrapper/file_lock.cpp [hunk 1]
```diff
@@ -60,6 +60,8 @@ cxx::expected<FileLockError> FileLock::initializeFileLock() noexcept
         return cxx::error<FileLockError>(convertErrnoToFileLockError(openCall.get_error().errnum));
     }
 
+    m_isInitialized = true;
```
Comment (qclzdh) [inline iceoryx_hoofs/source/posix_wrapper/file_lock.cpp:63]: It seems this change will cause the locked file removed unexpected.

Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/file_lock.cpp:63]: @qclzdh Good catch! You are right it should be closed only not removed - I fixed that.

**Review by elBoberido — DISMISSED** [review]

**Review by elBoberido — APPROVED** [review]

**Comment by elBoberido** [comment]
@dkroenke ping :)

**Pull request merged**